### PR TITLE
Add a new Device backed by flutter_tester.

### PR DESCRIPTION
**Pull request by scheglov**

### CODE REVIEW

File: packages/flutter_tools/lib/src/tester/flutter_tester.dart [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+        super(id: directory);
+
+  @override
+  String get name => fs.path.basename(_directory);
+
+  @override
+  String get packagePath => fs.path.join(_directory, '.packages');
+}
+
+// TODO(devoncarew): This device does not currently work with full restarts.
```
Comment (devoncarew) [inline packages/flutter_tools/lib/src/tester/flutter_tester.dart]: ==> `scheglov`

File: packages/flutter_tools/lib/src/tester/flutter_tester.dart [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+  Future<bool> get isLocalEmulator async => false;
+
+  @override
+  String get name => 'Flutter test device';
+
+  @override
+  DevicePortForwarder get portForwarder => null;
+
+  @override
+  Future<String> get sdkNameAndVersion async {
```
Comment (devoncarew) [inline packages/flutter_tools/lib/src/tester/flutter_tester.dart]: Does this need to be async? We could just return a Future.value().

File: packages/flutter_tools/lib/src/tester/flutter_tester.dart [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+  @override
+  Future<LaunchResult> startApp(ApplicationPackage package,
+      {@required String mainPath,
+      String route,
+      @required DebuggingOptions debuggingOptions,
+      Map<String, dynamic> platformArgs,
+      bool prebuiltApplication: false,
+      bool applicationNeedsRebuild: false,
+      bool usesTerminalUi: true,
+      bool ipv6: false}) async {
```
Comment (devoncarew) [inline packages/flutter_tools/lib/src/tester/flutter_tester.dart]: Could you add a trailing comma, for more better formatting?

**Review by devoncarew — COMMENTED** [review]
Some initial comments - have not yet had time to look at the tests.

File: packages/flutter_tools/lib/src/tester/flutter_tester.dart (hunk 1 above)
Comment (scheglov) [reply packages/flutter_tools/lib/src/tester/flutter_tester.dart]: Done.

File: packages/flutter_tools/lib/src/tester/flutter_tester.dart (hunk 2 above)
Comment (scheglov) [reply packages/flutter_tools/lib/src/tester/flutter_tester.dart]: Code with `async` is easier, and this is not a performance sensitive code.
So, it does not matter if (which I don't know) it is a tiny bit slower.
FWIW, other implementations of this method are also `async` without `await`.

File: packages/flutter_tools/lib/src/tester/flutter_tester.dart (hunk 3 above)
Comment (scheglov) [reply packages/flutter_tools/lib/src/tester/flutter_tester.dart]: Done.

File: packages/flutter_tools/lib/src/tester/flutter_tester.dart (hunk 2 above)
Comment (devoncarew) [reply packages/flutter_tools/lib/src/tester/flutter_tester.dart]: sgtm

**Review by devoncarew — APPROVED** [review]
lgtm!

For people following along, we plan to use this headless flutter device for CI integration testing for the Flutter IntelliJ plugin, and other similar use cases.

**Comment by Hixie** [comment]
What's the purpose of this device?

**Comment by devoncarew** [comment]
Three currently:
- it's used for integration testing flutter_tools, to catch regressions in hot reload and debugging (https://github.com/flutter/flutter/tree/master/packages/flutter_tools/test/integration)
- it's used for integration testing of IDEs, so they can run through full tests in CI systems w/o having to have native toolchains and emulators installed. VS Code in using it now, and we plan to do the same for IntelliJ
- we use it as a headless device to enable previewing individual widgets in the Flutter Outline view (https://medium.com/flutter-io/flutter-outline-hot-reload-and-the-implementation-of-a-live-widget-preview-69abd39aa3bb)

**Pull request merged**